src/util/mpir_hwtopo.c

```diff
@@ -11,53 +11,531 @@
+            hwloc_obj_type = HWLOC_OBJ_NUMANODE;
+            break;
+        case MPIR_HWTOPO_TYPE__PCI_DEVICE:
+            hwloc_obj_type = HWLOC_OBJ_PCI_DEVICE;
+            break;
+        default:
+            hwloc_obj_type = -1;
+    }
+
+    return hwloc_obj_type;
+}
+
+/* Get object type class in hwloc: Memory, I/O, Misc or Normal */
+static MPIR_hwtopo_class_e get_type_class(hwloc_obj_type_t type)
+{
+    MPIR_hwtopo_class_e class;
+
+    switch (type) {
+        case HWLOC_OBJ_NUMANODE:
+            class = MPIR_HWTOPO_CLASS__MEMORY;
+            break;
+        case HWLOC_OBJ_BRIDGE:
+        case HWLOC_OBJ_PCI_DEVICE:
+        case HWLOC_OBJ_OS_DEVICE:
+            class = MPIR_HWTOPO_CLASS__IO;
+            break;
+        case HWLOC_OBJ_MISC:
+            class = MPIR_HWTOPO_CLASS__MISC;
+            break;
+        case HWLOC_OBJ_MACHINE:
+        case HWLOC_OBJ_PACKAGE:
+        case HWLOC_OBJ_CORE:
+        case HWLOC_OBJ_PU:
+        case HWLOC_OBJ_L1CACHE:
+        case HWLOC_OBJ_L2CACHE:
+        case HWLOC_OBJ_L3CACHE:
+        case HWLOC_OBJ_L4CACHE:
+        case HWLOC_OBJ_L5CACHE:
+        case HWLOC_OBJ_L1ICACHE:
+        case HWLOC_OBJ_L2ICACHE:
+        case HWLOC_OBJ_L3ICACHE:
+        case HWLOC_OBJ_GROUP:
+            class = MPIR_HWTOPO_CLASS__NORMAL;
+            break;
+        default:
+            class = MPIR_HWTOPO_CLASS__INVALID;
+    }
+
+    return class;
+}
+#endif
+
+/*
+ * Hardware topology
+ */
+static struct {
+#ifdef HAVE_HWLOC
+    hwloc_topology_t hwloc_topology;
+    hwloc_cpuset_t bindset;
+#endif
+    int bindset_is_valid;
+} hwtopo = {
+0};
+
+struct shmem_processor_info_table {
+    const char *val;
+    MPIR_hwtopo_type_e type;
+};
+
+/* hwtopo node object table */
+static struct shmem_processor_info_table shmem_processor_info[] = {
+    {"node", MPIR_HWTOPO_TYPE__NODE},
+    {"machine", MPIR_HWTOPO_TYPE__NODE},
+    {"socket", MPIR_HWTOPO_TYPE__SOCKET},
+    {"package", MPIR_HWTOPO_TYPE__PACKAGE},
+    {"cpu", MPIR_HWTOPO_TYPE__CPU},
+    {"core", MPIR_HWTOPO_TYPE__CORE},
+    {"hwthread", MPIR_HWTOPO_TYPE__HWTHREAD},
+    {"pu", MPIR_HWTOPO_TYPE__HWTHREAD},
+    {"l1dcache", MPIR_HWTOPO_TYPE__L1CACHE},
+    {"l1ucache", MPIR_HWTOPO_TYPE__L1CACHE},
+    {"l1cache", MPIR_HWTOPO_TYPE__L1CACHE},
+    {"l2dcache", MPIR_HWTOPO_TYPE__L2CACHE},
+    {"l2ucache", MPIR_HWTOPO_TYPE__L2CACHE},
+    {"l2cache", MPIR_HWTOPO_TYPE__L2CACHE},
+    {"l3dcache", MPIR_HWTOPO_TYPE__L3CACHE},
+    {"l3ucache", MPIR_HWTOPO_TYPE__L3CACHE},
+    {"l3cache", MPIR_HWTOPO_TYPE__L3CACHE},
+    {"l4dcache", MPIR_HWTOPO_TYPE__L4CACHE},
+    {"l4ucache", MPIR_HWTOPO_TYPE__L4CACHE},
+    {"l4cache", MPIR_HWTOPO_TYPE__L4CACHE},
+    {"l5dcache", MPIR_HWTOPO_TYPE__L5CACHE},
+    {"l5ucache", MPIR_HWTOPO_TYPE__L5CACHE},
+    {"l5cache", MPIR_HWTOPO_TYPE__L5CACHE},
+    {"numanode", MPIR_HWTOPO_TYPE__DDR},
+    {"ddr", MPIR_HWTOPO_TYPE__DDR},
+    {"hbm", MPIR_HWTOPO_TYPE__HBM},
+    {NULL, MPIR_HWTOPO_TYPE__MAX}
+};
+
```

Comment:
This is only used for MPIR_hwtopo_get_type_id, right? Move it close to it.

Reply:
Actually, I think we could even move the static variable inside the function if it is only used by the function.

Reply:
Also, what is the significance of `shmem_processor` prefix?